In [6]:
!pip install pyspark

Defaulting to user installation because normal site-packages is not writeable
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 317.3/317.3 MB 120.5 MB/s eta 0:00:00m eta 0:00:010:01:01
  Preparing metadata (setup.py) ... done
  Created wheel for pyspark: filename=pyspark-3.5.3-py2.py3-none-any.whl size=317840629 sha256=72be29c711fb8f4a2d43c7ec7b772ec983eea6cebcd83713226f7d5845fbab05
  Stored in directory: /home/exouser/.cache/pip/wheels/07/a0/a3/d24c94bf043ab5c7e38c30491199a2a11fef8d2584e6df7fb7
Successfully built pyspark


In [2]:
import pyspark

In [3]:
from pyspark.sql import SparkSession

In [73]:
!pip install findspark

Defaulting to user installation because normal site-packages is not writeable


In [4]:
import findspark
findspark.init()

### 1

In [2]:
from pyspark.sql import SparkSession
spark = SparkSession.builder\
        .appName('Assignment2').\
        config( 'spark.executor.instances','2'). \
        config('spark.executor.cores','2'). \
        config('spark.executor-memory', '3g').\
        enableHiveSupport()\
        .getOrCreate()

Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
24/10/27 14:56:22 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


In [3]:
spark

In [5]:
spark._jsc.sc().getExecutorMemoryStatus().size()

1

24/10/27 14:56:41 WARN GarbageCollectionMetrics: To enable non-built-in garbage collector(s) List(G1 Concurrent GC), users should configure it(them) to spark.eventLog.gcMetrics.youngGenerationGarbageCollectors or spark.eventLog.gcMetrics.oldGenerationGarbageCollectors


In [6]:
df_cubzon= spark.read\
        .format('csv')\
        .option("header","true")\
        .option("inferschema","true")\
        .load("/home/exouser/bigdata/cubic_zirconia.csv")

In [7]:
df_cubzon.show()

+---+-----+---------+-----+-------+-----+-----+----+----+----+-----+
|_c0|carat|      cut|color|clarity|depth|table|   x|   y|   z|price|
+---+-----+---------+-----+-------+-----+-----+----+----+----+-----+
|  1|  0.3|    Ideal|    E|    SI1| 62.1| 58.0|4.27|4.29|2.66|  499|
|  2| 0.33|  Premium|    G|     IF| 60.8| 58.0|4.42|4.46| 2.7|  984|
|  3|  0.9|Very Good|    E|   VVS2| 62.2| 60.0|6.04|6.12|3.78| 6289|
|  4| 0.42|    Ideal|    F|    VS1| 61.6| 56.0|4.82| 4.8|2.96| 1082|
|  5| 0.31|    Ideal|    F|   VVS1| 60.4| 59.0|4.35|4.43|2.65|  779|
|  6| 1.02|    Ideal|    D|    VS2| 61.5| 56.0|6.46|6.49|3.99| 9502|
|  7| 1.01|     Good|    H|    SI1| 63.7| 60.0|6.35| 6.3|4.03| 4836|
|  8|  0.5|  Premium|    E|    SI1| 61.5| 62.0|5.09|5.06|3.12| 1415|
|  9| 1.21|     Good|    H|    SI1| 63.8| 64.0|6.72|6.63|4.26| 5407|
| 10| 0.35|    Ideal|    F|    VS2| 60.5| 57.0|4.52| 4.6|2.76|  706|
| 11| 0.32|    Ideal|    E|    VS2| 61.6| 56.0| 4.4|4.43|2.72|  637|
| 12|  1.1|  Premium|    D|    SI1

24/10/27 14:56:46 WARN CSVHeaderChecker: CSV header does not conform to the schema.
 Header: , carat, cut, color, clarity, depth, table, x, y, z, price
 Schema: _c0, carat, cut, color, clarity, depth, table, x, y, z, price
Expected: _c0 but found: 
CSV file: file:///home/exouser/bigdata/cubic_zirconia.csv


#### Part1

##### 1 Data Loading and Inspection

In [10]:
df_cubzon.show(5)

+---+-----+---------+-----+-------+-----+-----+----+----+----+-----+
|_c0|carat|      cut|color|clarity|depth|table|   x|   y|   z|price|
+---+-----+---------+-----+-------+-----+-----+----+----+----+-----+
|  1|  0.3|    Ideal|    E|    SI1| 62.1| 58.0|4.27|4.29|2.66|  499|
|  2| 0.33|  Premium|    G|     IF| 60.8| 58.0|4.42|4.46| 2.7|  984|
|  3|  0.9|Very Good|    E|   VVS2| 62.2| 60.0|6.04|6.12|3.78| 6289|
|  4| 0.42|    Ideal|    F|    VS1| 61.6| 56.0|4.82| 4.8|2.96| 1082|
|  5| 0.31|    Ideal|    F|   VVS1| 60.4| 59.0|4.35|4.43|2.65|  779|
+---+-----+---------+-----+-------+-----+-----+----+----+----+-----+
only showing top 5 rows



24/10/27 15:33:56 WARN CSVHeaderChecker: CSV header does not conform to the schema.
 Header: , carat, cut, color, clarity, depth, table, x, y, z, price
 Schema: _c0, carat, cut, color, clarity, depth, table, x, y, z, price
Expected: _c0 but found: 
CSV file: file:///home/exouser/bigdata/cubic_zirconia.csv


In [11]:
df_cubzon = df_cubzon.withColumnsRenamed({'carat' :'Carat', 'cut' :'Cut', 'color' :'Color' ,\
                             'clarity' :'Clarity','dept' :'Dept_percent','table':'Table_percent' \
                            , 'x' :'Length', 'y': 'Width','z': 'Height', 'price' :'Price'})
df_cubzon.show()

+---+-----+---------+-----+-------+-----+-------------+------+-----+------+-----+
|_c0|Carat|      Cut|Color|Clarity|depth|Table_percent|Length|Width|Height|Price|
+---+-----+---------+-----+-------+-----+-------------+------+-----+------+-----+
|  1|  0.3|    Ideal|    E|    SI1| 62.1|         58.0|  4.27| 4.29|  2.66|  499|
|  2| 0.33|  Premium|    G|     IF| 60.8|         58.0|  4.42| 4.46|   2.7|  984|
|  3|  0.9|Very Good|    E|   VVS2| 62.2|         60.0|  6.04| 6.12|  3.78| 6289|
|  4| 0.42|    Ideal|    F|    VS1| 61.6|         56.0|  4.82|  4.8|  2.96| 1082|
|  5| 0.31|    Ideal|    F|   VVS1| 60.4|         59.0|  4.35| 4.43|  2.65|  779|
|  6| 1.02|    Ideal|    D|    VS2| 61.5|         56.0|  6.46| 6.49|  3.99| 9502|
|  7| 1.01|     Good|    H|    SI1| 63.7|         60.0|  6.35|  6.3|  4.03| 4836|
|  8|  0.5|  Premium|    E|    SI1| 61.5|         62.0|  5.09| 5.06|  3.12| 1415|
|  9| 1.21|     Good|    H|    SI1| 63.8|         64.0|  6.72| 6.63|  4.26| 5407|
| 10| 0.35|    I

24/10/27 15:33:57 WARN CSVHeaderChecker: CSV header does not conform to the schema.
 Header: , carat, cut, color, clarity, depth, table, x, y, z, price
 Schema: _c0, carat, cut, color, clarity, depth, table, x, y, z, price
Expected: _c0 but found: 
CSV file: file:///home/exouser/bigdata/cubic_zirconia.csv


##### 2 Dataset Overview

In [12]:
print("number of rows ",df_cubzon.count(), "number of columns",len(df_cubzon.columns))

number of rows  26967 number of columns 11


In [13]:
# data types for each column
df_cubzon.printSchema()

root
 |-- _c0: integer (nullable = true)
 |-- Carat: double (nullable = true)
 |-- Cut: string (nullable = true)
 |-- Color: string (nullable = true)
 |-- Clarity: string (nullable = true)
 |-- depth: double (nullable = true)
 |-- Table_percent: double (nullable = true)
 |-- Length: double (nullable = true)
 |-- Width: double (nullable = true)
 |-- Height: double (nullable = true)
 |-- Price: integer (nullable = true)



In [14]:
#Unique values in Cut
df_cubzon.select('Cut').distinct().show()

+---------+
|      Cut|
+---------+
|  Premium|
|    Ideal|
|     Good|
|     Fair|
|Very Good|
+---------+



In [15]:
#Unique values in COLOR
df_cubzon.select('Color').distinct().show()

+-----+
|Color|
+-----+
|    F|
|    E|
|    D|
|    J|
|    G|
|    I|
|    H|
+-----+



In [16]:
#Unique values in Clarity
df_cubzon.select('Clarity').distinct().show()

+-------+
|Clarity|
+-------+
|   VVS2|
|    SI1|
|     IF|
|     I1|
|   VVS1|
|    VS2|
|    SI2|
|    VS1|
+-------+



#### Part 2

##### 1 Handling Missing Data:

In [17]:
##Missing values
from pyspark.sql.functions import isnan, when, count, col, format_number
nulls = df_cubzon.select([count(when(isnan(c) | col(c).isNull(), c)).alias(c) for c in df_cubzon.columns])
nulls.show()

24/10/27 15:35:04 WARN CSVHeaderChecker: CSV header does not conform to the schema.
 Header: , carat, cut, color, clarity, depth, table, x, y, z, price
 Schema: _c0, carat, cut, color, clarity, depth, table, x, y, z, price
Expected: _c0 but found: 
CSV file: file:///home/exouser/bigdata/cubic_zirconia.csv


+---+-----+---+-----+-------+-----+-------------+------+-----+------+-----+
|_c0|Carat|Cut|Color|Clarity|depth|Table_percent|Length|Width|Height|Price|
+---+-----+---+-----+-------+-----+-------------+------+-----+------+-----+
|  0|    0|  0|    0|      0|  697|            0|     0|    0|     0|    0|
+---+-----+---+-----+-------+-----+-------------+------+-----+------+-----+



This shows that depth has allot of null values

In [18]:
## lets check the percentage and handle missing values

percent = df_cubzon.select([format_number(((count(when(isnan(c) | col(c).isNull(), c))/df_cubzon.count())*100),1).alias(c) for c in df_cubzon.columns])
percent.show()

24/10/27 15:35:09 WARN CSVHeaderChecker: CSV header does not conform to the schema.
 Header: , carat, cut, color, clarity, depth, table, x, y, z, price
 Schema: _c0, carat, cut, color, clarity, depth, table, x, y, z, price
Expected: _c0 but found: 
CSV file: file:///home/exouser/bigdata/cubic_zirconia.csv


+---+-----+---+-----+-------+-----+-------------+------+-----+------+-----+
|_c0|Carat|Cut|Color|Clarity|depth|Table_percent|Length|Width|Height|Price|
+---+-----+---+-----+-------+-----+-------------+------+-----+------+-----+
|0.0|  0.0|0.0|  0.0|    0.0|  2.6|          0.0|   0.0|  0.0|   0.0|  0.0|
+---+-----+---+-----+-------+-----+-------------+------+-----+------+-----+



As the percent of missing values is very less, lets drop the missing values

In [19]:
df_cubzon_clean = df_cubzon.na.drop(subset = ["depth"])

In [20]:
df_cubzon_clean.count()

26270

##### 2. Descriptive statistics

In [32]:
## organising numerical and string columns
from pyspark.sql.types import *
num_ip = []
str_ip = []
for col in df_cubzon.columns:
    if str(df_cubzon_clean.schema[col].dataType) == 'StringType()':
        str_ip.append(col)
    else:
        num_ip.append(col)

In [33]:
num_ip

['_c0',
 'Carat',
 'depth',
 'Table_percent',
 'Length',
 'Width',
 'Height',
 'Price']

In [34]:
## remove index column
num_ip = num_ip[1:-1]
num_ip

['Carat', 'depth', 'Table_percent', 'Length', 'Width', 'Height']

In [35]:
str_ip

['Cut', 'Color', 'Clarity']

In [30]:
## summary statistics
df_cubzon_clean[num_ip].describe().show(truncate = False)

24/10/27 15:36:50 WARN SparkStringUtils: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.


+-------+-------------------+------------------+------------------+------------------+------------------+------------------+------------------+
|summary|Carat              |depth             |Table_percent     |Length            |Width             |Height            |Price             |
+-------+-------------------+------------------+------------------+------------------+------------------+------------------+------------------+
|count  |26270              |26270             |26270             |26270             |26270             |26270             |26270             |
|mean   |0.7979950513894071 |61.745146555006194|57.45601065854585 |5.7291674914351045|5.732967263037647 |3.53663608679101  |3937.9737342976778|
|stddev |0.47721435982056454|1.4128602381426116|2.2307763757680776|1.1275149511617892|1.1663302971035634|0.6993349197962537|4022.1893410338644|
|min    |0.2                |50.8              |49.0              |0.0               |0.0               |0.0               |326         

Looking at the data ,Carat and Price: With carat weights ranging from 0.2 to 4.5, combined with prices from $326$ up to nearly $19,000$, there’s a likely relationship between size and price, though the variability in price suggests that other factors (like clarity, cut quality, and rarity) are at play as well.

In [36]:
#### outliers
from pyspark.sql.functions import *
from pyspark.ml.feature import StringIndexer, VectorAssembler
d = {}
for col in num_ip:
    d[col] = df_cubzon_clean.approxQuantile(col,[0.25,0.50,0.75,0.99],0.25)

d

{'Carat': [0.2, 0.4, 4.5, 4.5],
 'depth': [50.8, 61.0, 73.6, 73.6],
 'Table_percent': [49.0, 56.0, 79.0, 79.0],
 'Length': [0.0, 4.71, 10.23, 10.23],
 'Width': [0.0, 4.72, 58.9, 58.9],
 'Height': [0.0, 2.9, 8.06, 8.06]}

In [37]:
df = df_cubzon_clean
for col in num_ip:
    skew = df.agg(skewness(df[col])).collect()
    skew = skew[0][0]
    if skew > 1:
        df = df.withColumn(col, \
        log(when(df[col]< d[col][0],d[col][0])\
        .when(df[col]>d[col][1],d[col][1])\
        .otherwise(df[col]) + 1)).alias(col)
        print(col+" has been treated for positive (right) skewness. (skew =",skew,")")
    elif skew < -1:
        df = df.withColumn(col, \
        exp(when(df[col]< d[col][0],d[col][0])\
        .when(df[col]>d[col][1],d[col][1])\
        .otherwise(df[col]) + 1)).alias(col)
        print(col+" has been treated for negative (left) skewness. (skew =",skew,")")
            

Carat has been treated for positive (right) skewness. (skew = 1.1176883595677425 )
Width has been treated for positive (right) skewness. (skew = 3.9392923373162962 )


##### 3.Categorical Value Distribution:

In [38]:
from pyspark.sql import functions as F
def dist_cal(col):
    dist = df_cubzon_clean.groupBy(col).count()
    total_count = df_cubzon_clean.count()
    dist = dist.withColumn("dist_percent",(F.col("count")/total_count) * 100)
    return dist

In [39]:
for col in str_ip:
    print("distribution for "+col)
    dist_cal(col).show()
    

distribution for Cut
+---------+-----+------------------+
|      Cut|count|      dist_percent|
+---------+-----+------------------+
|  Premium| 6707|25.531023981728207|
|    Ideal|10546|40.144651693947466|
|     Good| 2382| 9.067377236391321|
|     Fair|  757| 2.881614008374572|
|Very Good| 5878| 22.37533307955843|
+---------+-----+------------------+

distribution for Color
+-----+-----+------------------+
|Color|count|      dist_percent|
+-----+-----+------------------+
|    F| 4612| 17.55614769699277|
|    E| 4793| 18.24514655500571|
|    D| 3268|  12.4400456794823|
|    J| 1401| 5.333079558431671|
|    G| 5529| 21.04682146935668|
|    I| 2676|10.186524552721735|
|    H| 3991|15.192234488009134|
+-----+-----+------------------+

distribution for Clarity
+-------+-----+------------------+
|Clarity|count|      dist_percent|
+-------+-----+------------------+
|   VVS2| 2479|  9.43661971830986|
|    SI1| 6408|24.392843547773126|
|     IF|  874|3.3269889607917777|
|     I1|  355|1.351351

#### Part3

##### 1. Feature engineering

In [40]:
# Feature Engineering
df_cubzon_clean = df_cubzon_clean.withColumn("Volume",(df_cubzon_clean["Length"] * df_cubzon_clean["Width"] * df_cubzon_clean["Height"]))

In [41]:
df_cubzon_clean.show()

+---+-----+---------+-----+-------+-----+-------------+------+-----+------+-----+------------------+
|_c0|Carat|      Cut|Color|Clarity|depth|Table_percent|Length|Width|Height|Price|            Volume|
+---+-----+---------+-----+-------+-----+-------------+------+-----+------+-----+------------------+
|  1|  0.3|    Ideal|    E|    SI1| 62.1|         58.0|  4.27| 4.29|  2.66|  499| 48.72667799999999|
|  2| 0.33|  Premium|    G|     IF| 60.8|         58.0|  4.42| 4.46|   2.7|  984|53.225640000000006|
|  3|  0.9|Very Good|    E|   VVS2| 62.2|         60.0|  6.04| 6.12|  3.78| 6289|        139.726944|
|  4| 0.42|    Ideal|    F|    VS1| 61.6|         56.0|  4.82|  4.8|  2.96| 1082| 68.48255999999999|
|  5| 0.31|    Ideal|    F|   VVS1| 60.4|         59.0|  4.35| 4.43|  2.65|  779|51.066824999999994|
|  6| 1.02|    Ideal|    D|    VS2| 61.5|         56.0|  6.46| 6.49|  3.99| 9502|167.28234600000002|
|  7| 1.01|     Good|    H|    SI1| 63.7|         60.0|  6.35|  6.3|  4.03| 4836|         1

24/10/27 15:52:22 WARN CSVHeaderChecker: CSV header does not conform to the schema.
 Header: , carat, cut, color, clarity, depth, table, x, y, z, price
 Schema: _c0, carat, cut, color, clarity, depth, table, x, y, z, price
Expected: _c0 but found: 
CSV file: file:///home/exouser/bigdata/cubic_zirconia.csv


For diamonds, Volume is the best feature to predict the price, where as carat is the size of the diamond, volume and carat plays major role in defining the price of diamond.

##### 2.Feature Scaling


In [43]:
# lets clean the input outputs and generate features columns / vector
num_ip.remove('Length')
num_ip.remove('Width')
num_ip.remove('Height')
num_ip.append('Volume')
num_ip

['Carat', 'depth', 'Table_percent', 'Volume']

In [44]:
#Feature Vector
assembler = VectorAssembler(inputCols=num_ip, outputCol= 'features')
output = assembler.transform(df_cubzon_clean)


Based on the features we have , Normalization is not required, as it scales b/w the range(0,1], this is used when we want to preserve zero. Where as standardiastion helps to center the data around the mean with standard deviation 1

In [45]:
from pyspark.ml.feature import MinMaxScaler, StandardScaler
scaler = StandardScaler(inputCol= 'features' , outputCol = 'standarad_features')
standard_model = scaler.fit(output)
std_data = standard_model.transform(output)

In [46]:
std_data.show(5)

+---+-----+---------+-----+-------+-----+-------------+------+-----+------+-----+------------------+--------------------+--------------------+
|_c0|Carat|      Cut|Color|Clarity|depth|Table_percent|Length|Width|Height|Price|            Volume|            features|  standarad_features|
+---+-----+---------+-----+-------+-----+-------------+------+-----+------+-----+------------------+--------------------+--------------------+
|  1|  0.3|    Ideal|    E|    SI1| 62.1|         58.0|  4.27| 4.29|  2.66|  499| 48.72667799999999|[0.3,62.1,58.0,48...|[0.62864830830489...|
|  2| 0.33|  Premium|    G|     IF| 60.8|         58.0|  4.42| 4.46|   2.7|  984|53.225640000000006|[0.33,60.8,58.0,5...|[0.69151313913538...|
|  3|  0.9|Very Good|    E|   VVS2| 62.2|         60.0|  6.04| 6.12|  3.78| 6289|        139.726944|[0.9,62.2,60.0,13...|[1.88594492491467...|
|  4| 0.42|    Ideal|    F|    VS1| 61.6|         56.0|  4.82|  4.8|  2.96| 1082| 68.48255999999999|[0.42,61.6,56.0,6...|[0.88010763162684...|

24/10/27 15:54:47 WARN CSVHeaderChecker: CSV header does not conform to the schema.
 Header: , carat, cut, color, clarity, depth, table, x, y, z, price
 Schema: _c0, carat, cut, color, clarity, depth, table, x, y, z, price
Expected: _c0 but found: 
CSV file: file:///home/exouser/bigdata/cubic_zirconia.csv


##### 3.Relationship Between Features

In [50]:
corr = {}
for col in num_ip:
    corr[col] = df_cubzon_clean.stat.corr(col,'Price')
    

In [51]:
corr

{'Carat': 0.9221335932129359,
 'depth': -0.002568616329956218,
 'Table_percent': 0.1265870119852145,
 'Volume': 0.8878757976746712}

Focusing on features such a Carat and Volume as it has a strong positive correlation and dept is negligible as it is close to zero and table_percent is weak positive correlation.

In [98]:
std_data.show(5) #our latest dataframe

+---+-----+---------+-----+-------+-----+-------------+------+-----+------+-----+------------------+--------------------+--------------------+
|_c0|Carat|      Cut|Color|Clarity|depth|Table_percent|Length|Width|Height|Price|            Volume|            features|  standarad_features|
+---+-----+---------+-----+-------+-----+-------------+------+-----+------+-----+------------------+--------------------+--------------------+
|  1|  0.3|    Ideal|    E|    SI1| 62.1|         58.0|  4.27| 4.29|  2.66|  499| 48.72667799999999|[0.3,62.1,58.0,48...|[0.62864830830489...|
|  2| 0.33|  Premium|    G|     IF| 60.8|         58.0|  4.42| 4.46|   2.7|  984|53.225640000000006|[0.33,60.8,58.0,5...|[0.69151313913538...|
|  3|  0.9|Very Good|    E|   VVS2| 62.2|         60.0|  6.04| 6.12|  3.78| 6289|        139.726944|[0.9,62.2,60.0,13...|[1.88594492491467...|
|  4| 0.42|    Ideal|    F|    VS1| 61.6|         56.0|  4.82|  4.8|  2.96| 1082| 68.48255999999999|[0.42,61.6,56.0,6...|[0.88010763162684...|

24/10/27 16:15:10 WARN CSVHeaderChecker: CSV header does not conform to the schema.
 Header: , carat, cut, color, clarity, depth, table, x, y, z, price
 Schema: _c0, carat, cut, color, clarity, depth, table, x, y, z, price
Expected: _c0 but found: 
CSV file: file:///home/exouser/bigdata/cubic_zirconia.csv


#### Part 4

##### 1.Data Preprocessing for Modeling

In [99]:
#converting categorical to numerical
columns = num_ip+str_ip + ['Price']
df = std_data.select(columns)
if len(str_ip)!=0:
    for col in str_ip:
        indexer = StringIndexer(inputCol= col , outputCol=col+"_num")
        df  = indexer.fit(df).transform(df)
        num_ip.append(col+"_num")

In [100]:
df.show(5)

+-----+-----+-------------+------------------+---------+-----+-------+-----+-------+---------+-----------+
|Carat|depth|Table_percent|            Volume|      Cut|Color|Clarity|Price|Cut_num|Color_num|Clarity_num|
+-----+-----+-------------+------------------+---------+-----+-------+-----+-------+---------+-----------+
|  0.3| 62.1|         58.0| 48.72667799999999|    Ideal|    E|    SI1|  499|    0.0|      1.0|        0.0|
| 0.33| 60.8|         58.0|53.225640000000006|  Premium|    G|     IF|  984|    1.0|      0.0|        6.0|
|  0.9| 62.2|         60.0|        139.726944|Very Good|    E|   VVS2| 6289|    2.0|      1.0|        4.0|
| 0.42| 61.6|         56.0| 68.48255999999999|    Ideal|    F|    VS1| 1082|    0.0|      2.0|        3.0|
| 0.31| 60.4|         59.0|51.066824999999994|    Ideal|    F|   VVS1|  779|    0.0|      2.0|        5.0|
+-----+-----+-------------+------------------+---------+-----+-------+-----+-------+---------+-----------+
only showing top 5 rows



In [101]:
features = num_ip
features

['Carat',
 'depth',
 'Table_percent',
 'Volume',
 'Cut_num',
 'Color_num',
 'Clarity_num']

In [102]:
final_columns = features + ['Price']

In [103]:
## Lets create a feature vector considering both the categorical(numerical converted data ) and existing numerical data, if required let's scale
df.select(final_columns).show()

+-----+-----+-------------+------------------+-------+---------+-----------+-----+
|Carat|depth|Table_percent|            Volume|Cut_num|Color_num|Clarity_num|Price|
+-----+-----+-------------+------------------+-------+---------+-----------+-----+
|  0.3| 62.1|         58.0| 48.72667799999999|    0.0|      1.0|        0.0|  499|
| 0.33| 60.8|         58.0|53.225640000000006|    1.0|      0.0|        6.0|  984|
|  0.9| 62.2|         60.0|        139.726944|    2.0|      1.0|        4.0| 6289|
| 0.42| 61.6|         56.0| 68.48255999999999|    0.0|      2.0|        3.0| 1082|
| 0.31| 60.4|         59.0|51.066824999999994|    0.0|      2.0|        5.0|  779|
| 1.02| 61.5|         56.0|167.28234600000002|    0.0|      4.0|        1.0| 9502|
| 1.01| 63.7|         60.0|         161.22015|    3.0|      3.0|        0.0| 4836|
|  0.5| 61.5|         62.0|         80.356848|    1.0|      1.0|        0.0| 1415|
| 1.21| 63.8|         64.0|189.79833599999998|    3.0|      3.0|        0.0| 5407|
| 0.

In [104]:
#Feature Vector
assembler = VectorAssembler(inputCols=features, outputCol= 'features')
df_vect = assembler.transform(df)
scaler = StandardScaler(inputCol= 'features' , outputCol = 'standarad_features')
standard_model = scaler.fit(df_vect)
df_std = standard_model.transform(df_vect)

+-----+-----+-------------+------------------+---------+-----+-------+-----+-------+---------+-----------+--------------------+--------------------+
|Carat|depth|Table_percent|            Volume|      Cut|Color|Clarity|Price|Cut_num|Color_num|Clarity_num|            features|  standarad_features|
+-----+-----+-------------+------------------+---------+-----+-------+-----+-------+---------+-----------+--------------------+--------------------+
|  0.3| 62.1|         58.0| 48.72667799999999|    Ideal|    E|    SI1|  499|    0.0|      1.0|        0.0|[0.3,62.1,58.0,48...|[0.62864830830489...|
| 0.33| 60.8|         58.0|53.225640000000006|  Premium|    G|     IF|  984|    1.0|      0.0|        6.0|[0.33,60.8,58.0,5...|[0.69151313913538...|
|  0.9| 62.2|         60.0|        139.726944|Very Good|    E|   VVS2| 6289|    2.0|      1.0|        4.0|[0.9,62.2,60.0,13...|[1.88594492491467...|
| 0.42| 61.6|         56.0| 68.48255999999999|    Ideal|    F|    VS1| 1082|    0.0|      2.0|        3.0|

In [106]:

df_final = df_std.select(['standarad_features','Price'])
df_final = df_final.withColumnRenamed('standarad_features','features')
df_final.show(5)

+--------------------+-----+
|            features|Price|
+--------------------+-----+
|[0.62864830830489...|  499|
|[0.69151313913538...|  984|
|[1.88594492491467...| 6289|
|[0.88010763162684...| 1082|
|[0.64960325191505...|  779|
+--------------------+-----+
only showing top 5 rows



In [107]:
### equal distribution of Price
quantiles = df_final.approxQuantile("Price" , [0.25,0.5,0.75],0)
bins = [float('-inf')] + quantiles + [float('inf')]
print('bins ' , bins)
df_stratified = df_final.withColumn("Price_bins", F.when(F.col('Price')<bins[1],0).when((F.col('Price')>bins[1]) & (F.col('Price')<bins[2]),1) \
                            .when((F.col('Price')>bins[2]) & (F.col('Price')<bins[3]),2).otherwise(3))

bins  [-inf, 945.0, 2375.0, 5361.0, inf]


In [108]:
df_stratified.show()

+--------------------+-----+----------+
|            features|Price|Price_bins|
+--------------------+-----+----------+
|[0.62864830830489...|  499|         0|
|[0.69151313913538...|  984|         1|
|[1.88594492491467...| 6289|         3|
|[0.88010763162684...| 1082|         1|
|[0.64960325191505...|  779|         0|
|[2.13740424823663...| 9502|         3|
|[2.11644930462647...| 4836|         2|
|[1.04774718050815...| 1415|         1|
|[2.53554817682973...| 5407|         3|
|[0.73342302635570...|  706|         0|
|[0.67055819552521...|  637|         0|
|[2.30504379711793...| 6468|         3|
|[1.04774718050815...| 1932|         1|
|[1.48780099632157...| 2767|         2|
|[3.14324154152446...|10644|         3|
|[0.64960325191505...|  544|         0|
|[0.71246808274554...|  650|         0|
|[2.11644930462647...| 7127|         3|
|[1.88594492491467...| 3567|         2|
|[1.13156695494880...| 1637|         1|
+--------------------+-----+----------+
only showing top 20 rows



In [123]:
train,test = df_stratified.randomSplit([0.8,0.2],seed= 1)

In [124]:
train.agg(F.mean('Price')).collect()

[Row(avg(Price)=3937.0982202341297)]

In [125]:
test.agg(F.mean('Price')).collect()

[Row(avg(Price)=3941.474124809741)]

both train and test has approximate mean 

##### 2. Regression model

In [156]:
from pyspark.ml.regression import LinearRegression, DecisionTreeRegressor
from pyspark.ml.evaluation import RegressionEvaluator

In [159]:
lr = LinearRegression(labelCol="Price", featuresCol="features")
lr_model = lr.fit(train)
lr_predictions = lr_model.transform(test)

24/10/27 16:46:28 WARN Instrumentation: [3f33f639] regParam is zero, which might cause numerical instability and overfitting.


In [160]:
dt = DecisionTreeRegressor(labelCol="Price", featuresCol="features")
dt_model = dt.fit(train)
dt_predictions = dt_model.transform(test)

##### 3. Model Evaluation

In [162]:
evaluator_rmse = RegressionEvaluator(labelCol="Price", predictionCol="prediction", metricName="rmse")
evaluator_mae = RegressionEvaluator(labelCol="Price", predictionCol="prediction", metricName="mae")
evaluator_r2 = RegressionEvaluator(labelCol="Price", predictionCol="prediction", metricName="r2")

# Linear Regression Metrics
lr_rmse = evaluator_rmse.evaluate(lr_predictions)
lr_mae = evaluator_mae.evaluate(lr_predictions)
lr_r2 = evaluator_r2.evaluate(lr_predictions)

# Decision Tree Metrics
dt_rmse = evaluator_rmse.evaluate(dt_predictions)
dt_mae = evaluator_mae.evaluate(dt_predictions)
dt_r2 = evaluator_r2.evaluate(dt_predictions)

# Display results
print("Linear Regression - RMSE:", lr_rmse, "MAE:", lr_mae, "R2:", lr_r2)
print("Decision Tree - RMSE:", dt_rmse, "MAE:", dt_mae, "R2:", dt_r2)

Linear Regression - RMSE: 1519.1445460268512 MAE: 988.3031741653074 R2: 0.8575301806811857
Decision Tree - RMSE: 1135.4104504536863 MAE: 620.3638467487106 R2: 0.9204151411476595


Decision Tree Performs better here as we have included both the numerical and numerical converted categorical columns as a feature.

####  Bonus

In [140]:
from pyspark.ml.tuning import CrossValidator, ParamGridBuilder
regressor = DecisionTreeRegressor(labelCol='Price', predictionCol="prediction")

# Build your parameter grid
paramGrid = (ParamGridBuilder() \
                         .addGrid(regressor.maxDepth, [2, 5, 10, 20, 30]) \
                         .addGrid(regressor.maxBins, [10, 20, 40]) \
                         .build())

#Cross Validator requires all of the following parameters:
crossval = CrossValidator(estimator=regressor,
                          estimatorParamMaps=paramGrid,
                          evaluator=RegressionEvaluator(labelCol='Price', predictionCol="prediction",metricName="rmse"),
                          numFolds=2)

# Fit Model: Run cross-validation, and choose the best set of parameters.
fitModel = crossval.fit(train)

# Get Best Model
DT_BestModel = fitModel.bestModel

print(" ")
print('\033[1m' + " Feature Importances"+ '\033[0m')
print("(Scores add up to 1)")
print("Lowest score is the least important")
print(" ")
DT_featureImportances = DT_BestModel.featureImportances.toArray()
# Convert from numpy array to list
imp_scores = []
for x in DT_featureImportances:
    imp_scores.append(float(x))
# Then zip with input_columns list and create a df
result = spark.createDataFrame(zip(features,imp_scores), schema=['feature','score'])
print(result.orderBy(result["score"].desc()).show(truncate=False))

# Make predictions.
# PySpark will automatically use the best model when you call fitmodel
predictions = fitModel.transform(test)


eval = RegressionEvaluator(labelCol='Price', predictionCol="prediction",metricName="rmse")
# And then apply it your predictions dataframe
rmse = eval.evaluate(predictions)
print(rmse)

24/10/27 16:33:09 WARN CacheManager: Asked to cache already cached data.
24/10/27 16:33:09 WARN CacheManager: Asked to cache already cached data.
24/10/27 16:33:13 WARN DAGScheduler: Broadcasting large task binary with size 1014.4 KiB
24/10/27 16:33:13 WARN DAGScheduler: Broadcasting large task binary with size 1267.7 KiB
24/10/27 16:33:14 WARN DAGScheduler: Broadcasting large task binary with size 1531.0 KiB
24/10/27 16:33:14 WARN DAGScheduler: Broadcasting large task binary with size 1771.3 KiB
24/10/27 16:33:14 WARN DAGScheduler: Broadcasting large task binary with size 1975.3 KiB
24/10/27 16:33:14 WARN DAGScheduler: Broadcasting large task binary with size 2.1 MiB
24/10/27 16:33:14 WARN DAGScheduler: Broadcasting large task binary with size 1347.8 KiB
24/10/27 16:33:15 WARN DAGScheduler: Broadcasting large task binary with size 1063.2 KiB
24/10/27 16:33:15 WARN DAGScheduler: Broadcasting large task binary with size 1324.4 KiB
24/10/27 16:33:15 WARN DAGScheduler: Broadcasting large 

 
 Feature Importances
(Scores add up to 1)
Lowest score is the least important
 


+-------------+---------------------+
|feature      |score                |
+-------------+---------------------+
|Carat        |0.6743371788214425   |
|Volume       |0.22642435309812284  |
|Clarity_num  |0.0671443951509537   |
|Color_num    |0.027835290117624933 |
|depth        |0.0015998489023920832|
|Cut_num      |0.001411161933810618 |
|Table_percent|0.0012477719756532374|
+-------------+---------------------+

None
759.4924163111617


From the above it is clear that the Decision Treee regression performs better in predicting the prices.

#### Trail 
Lets remove categorical values (numerical converted) - 
try the same with removing categorical columns, and check model performance

In [178]:
#data preprocessing and modelling
# features = features[:-3]
#Feature Vector
assembler = VectorAssembler(inputCols=features, outputCol= 'features')
df_vect = assembler.transform(df)
#standardisation
scaler = StandardScaler(inputCol= 'features' , outputCol = 'standarad_features')
standard_model = scaler.fit(df_vect)
df_std = standard_model.transform(df_vect)
df_final = df_std.select(['standarad_features','Price'])
df_final = df_final.withColumnRenamed('standarad_features','features')
df_final.show(5,truncate = False)

+-----------------------------------------------------------------------------+-----+
|features                                                                     |Price|
+-----------------------------------------------------------------------------+-----+
|[0.628648308304892,43.953392079062624,25.99991672407327,0.6067842242318457]  |499  |
|[0.6915131391353814,43.03327276017725,25.99991672407327,0.6628089580956762]  |984  |
|[1.8859449249146765,44.02417048820765,26.89646557662752,1.7399935476686215]  |6289 |
|[0.8801076316268489,43.59950003333748,25.10336787151902,0.8528005345041342]  |1082 |
|[0.6496032519150552,42.750159123597136,26.448191150350393,0.6359256379351046]|779  |
+-----------------------------------------------------------------------------+-----+
only showing top 5 rows



In [179]:
### equal distribution of Price
quantiles = df_final.approxQuantile("Price" , [0.25,0.5,0.75],0)
bins = [float('-inf')] + quantiles + [float('inf')]
print('bins ' , bins)
df_stratified = df_final.withColumn("Price_bins", F.when(F.col('Price')<bins[1],0).when((F.col('Price')>bins[1]) & (F.col('Price')<bins[2]),1) \
                            .when((F.col('Price')>bins[2]) & (F.col('Price')<bins[3]),2).otherwise(3))
train,test = df_stratified.randomSplit([0.8,0.2],seed= 1)
train.agg(F.mean('Price')).collect()

bins  [-inf, 945.0, 2375.0, 5361.0, inf]


[Row(avg(Price)=3935.6397639668794)]

In [180]:
test.agg(F.mean('Price')).collect() # both means should be approximately same

[Row(avg(Price)=3947.305175038052)]

In [181]:
#Linear regression and Decision tree comparision
lr = LinearRegression(labelCol="Price", featuresCol="features", regParam=0.5)
lr_model = lr.fit(train)
lr_predictions = lr_model.transform(test)
dt = DecisionTreeRegressor(labelCol="Price", featuresCol="features")
dt_model = dt.fit(train)
dt_predictions = dt_model.transform(test)
evaluator_rmse = RegressionEvaluator(labelCol="Price", predictionCol="prediction", metricName="rmse")
evaluator_mae = RegressionEvaluator(labelCol="Price", predictionCol="prediction", metricName="mae")
evaluator_r2 = RegressionEvaluator(labelCol="Price", predictionCol="prediction", metricName="r2")

# Linear Regression Metrics
lr_rmse = evaluator_rmse.evaluate(lr_predictions)
lr_mae = evaluator_mae.evaluate(lr_predictions)
lr_r2 = evaluator_r2.evaluate(lr_predictions)

# Decision Tree Metrics
dt_rmse = evaluator_rmse.evaluate(dt_predictions)
dt_mae = evaluator_mae.evaluate(dt_predictions)
dt_r2 = evaluator_r2.evaluate(dt_predictions)

# Display results
print("Linear Regression - RMSE:", lr_rmse, "MAE:", lr_mae, "R2:", lr_r2)
print("Decision Tree - RMSE:", dt_rmse, "MAE:", dt_mae, "R2:", dt_r2)

Linear Regression - RMSE: 1577.4124255136417 MAE: 1027.8491628805155 R2: 0.8466236723058271
Decision Tree - RMSE: 1462.2618623929739 MAE: 828.8615641833175 R2: 0.8681991756426296


In [182]:
## Decision Tree performs well